subprojects/execution/src/main/java/org/gradle/internal/execution/steps/ExecuteStep.java

```diff
@@ -57,16 +47,4 @@ private ExecutionOutcome determineOutcome(UnitOfWork.WorkResult result, boolean
                 throw new IllegalArgumentException("Unknown result: " + result);
         }
     }
-
-    private InputChangesInternal determineInputChanges(UnitOfWork work, IncrementalChangesContext context) {
-        return context.getChanges()
-            .map(changes -> {
-                InputChangesInternal inputChanges = changes.createInputChanges();
-                if (!inputChanges.isIncremental()) {
-                    LOGGER.info("All input files are considered out-of-date for incremental {}.", work.getDisplayName());
-                }
-                return inputChanges;
-            })
-        .orElseThrow(() -> new UnsupportedOperationException("Cannot use input changes when input tracking is disabled."));
-    }
```

Comment:
To me this would look better, and wouldn't require `InputChangesContext.isIncrementalExecution()`:

```java
public class ExecuteStep<C extends InputChangesContext> implements Step<C, Result> {

    @Override
    public Result execute(C context) {
        UnitOfWork work = context.getWork();
        ExecutionOutcome outcome = context.getInputChanges()
            .map(inputChanges -> determineOutcome(work.execute(inputChanges), EXECUTED_INCREMENTALLY))
            .orElseGet(() -> determineOutcome(work.execute(null), EXECUTED_NON_INCREMENTALLY));
        return new Result() {
            @Override
            public Try<ExecutionOutcome> getOutcome() {
                return Try.successful(outcome);
            }
        };
    }

    private static ExecutionOutcome determineOutcome(WorkResult result, ExecutionOutcome didWorkOutcome) {
        switch (result) {
            case DID_NO_WORK:
                return ExecutionOutcome.UP_TO_DATE;
            case DID_WORK:
                return didWorkOutcome;
            default:
                throw new AssertionError();
        }
    }
}
```

Reply:
Changed with a slight change: If there are input changes, we still need to check whether the input changes require a rebuild or not.

Reply:
👍